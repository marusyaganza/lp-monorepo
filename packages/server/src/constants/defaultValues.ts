import { Language, Role, SortWordsBy, Tense } from '../generated/graphql';

export const WORDS_PER_PAGE = 5;
export const DEFAULT_LANGUAGE = Language.English;
export const DEFAULT_WORDS_SORT_BY = SortWordsBy.UpdatedAt;
export const DEFAULT_ROLE = Role.Member;
export const DEFAULT_TENSE = Tense.Pind;
export const DEFAULT_TOKEN_TTL = 7;
// to clean db every hour + 10 min just in case
export const DEMO_DB_TTL = 4200;
export const DEFAULT_MONGO_CONNECTION = 'mongodb://localhost:27017';
export const DEFAULT_OPTIONS__SAMPLE_SIZE = 50;
