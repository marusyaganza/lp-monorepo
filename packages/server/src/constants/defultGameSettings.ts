import { Game, Language, SettingsByGame } from '../generated/graphql';

export const DEFAULT_GAMES_SETTINGS: SettingsByGame = {
  [Game.Audio]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Image]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Speaking]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.TypeWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectDef]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Conjugation]: {
    wordsPerGame: 2,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  },
  [Game.Gender]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  }
};
