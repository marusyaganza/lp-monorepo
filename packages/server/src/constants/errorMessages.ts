export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
