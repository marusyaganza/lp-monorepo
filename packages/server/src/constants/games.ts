import { Game, GameConfig, Language } from '../generated/graphql';
export const GAMES: GameConfig[] = [
  {
    name: 'WordEcho',
    id: 'audio',
    desc: 'Listen, Type, Master. Sharpen language skills through auditory engagement.',
    type: Game.Audio,
    orderNum: 0,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'LexiSense',
    desc: 'Define & Conquer.Match words with their meanings intuitively.',
    id: 'select_definition',
    type: Game.SelectDef,
    orderNum: 1,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Definify: Match & Learn',
    desc: 'Associate definitions with the right words. A captivating word puzzle adventure!',
    id: 'select_word',
    type: Game.SelectWord,
    orderNum: 2,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'WordRecall: Define & Type',
    desc: 'Type words matching given definitions. Boost vocabulary in engaging rounds!',
    id: 'type_word',
    type: Game.TypeWord,
    orderNum: 3,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'PictureWord Blitz',
    desc: 'Identify objects from pictures and type their names to enhance vocabulary skills!',
    id: 'image',
    type: Game.Image,
    orderNum: 4,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'VoiceWord Challenge',
    desc: 'Speak words from clues and images. Sharpen your vocabulary and speaking skills fast!',
    id: 'speaking',
    type: Game.Speaking,
    orderNum: 5,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Conjugation Quest',
    desc: 'Master verb conjugations across tenses and languages in fast-paced, interactive challenges!',
    id: 'conjugation',
    type: Game.Conjugation,
    orderNum: 6,
    languages: [Language.Spanish]
  },
  {
    name: 'GenderWord Challenge',
    desc: 'Choose nouns and match their gender. Test linguistic skills in thrilling rounds!',
    id: 'gender',
    type: Game.Gender,
    orderNum: 7,
    languages: [Language.Spanish]
  }
];
