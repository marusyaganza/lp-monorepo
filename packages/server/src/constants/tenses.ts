import { Tense } from '../generated/graphql';

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto',
  futr: 'Futuro indicativo',
  cond: 'Condicional',
  psub: 'Presente subjentivo'
  // gppt: 'Gerundio'
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
