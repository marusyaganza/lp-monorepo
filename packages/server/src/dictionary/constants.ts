import { Tense } from 'generated/graphql';

export type TagType = {
  tag: string;
  replacement: string;
};

export type ComplexTagType = {
  opening: string;
  closing: string;
};

export const TAGS: TagType[] = [
  { tag: '{b}', replacement: '<b>' },
  { tag: '{/b}', replacement: '</b>' },
  { tag: '{ldquo}', replacement: '\u201C' },
  { tag: '{rdquo}', replacement: '\u201D' },
  { tag: '{it}', replacement: '<i>' },
  { tag: '{/it}', replacement: '</i>' },
  { tag: '{wi}', replacement: '<i>' },
  { tag: '{/wi}', replacement: '</i>' },
  { tag: '{gloss}=', replacement: '<i>' },
  { tag: '{/gloss}', replacement: '</i>' },
  { tag: '{sc}', replacement: '<i>' },
  { tag: '{/sc}', replacement: '</i>' },
  { tag: '{parahw}', replacement: '</i>' },
  { tag: '{/parahw}', replacement: '</i>' },
  { tag: '{phrase}', replacement: '</i>' },
  { tag: '{/phrase}', replacement: '</i>' },
  { tag: '{qword}', replacement: '</i>' },
  { tag: '{/qword}', replacement: '</i>' },
  { tag: '{itsc}', replacement: '<i>' },
  { tag: '{/itsc}', replacement: '</i>' },
  { tag: '{inf}', replacement: '' },
  { tag: '{/inf}', replacement: '' },
  { tag: '{sup}', replacement: '' },
  { tag: '{/sup}', replacement: '' },
  { tag: '{bc}', replacement: '' },
  { tag: '{dx}', replacement: '' },
  { tag: '{/dx}', replacement: '' },
  { tag: '{dx_def}', replacement: '' },
  { tag: '{/dx_def}', replacement: '' },
  { tag: '{dx_ety}', replacement: '' },
  { tag: '{/dx_ety}', replacement: '' },
  { tag: '{ma}', replacement: '' },
  { tag: '{/ma}', replacement: '' },
  { tag: '{rom}', replacement: '' },
  { tag: '{/rom}', replacement: '' }
];

export const COMPLEX_TAGS: ComplexTagType[] = [
  { opening: '{sx', closing: '}' },
  { opening: '{d_link', closing: '}' },
  { opening: '{a_link', closing: '}' },
  { opening: '{i_link', closing: '}' },
  { opening: '{et_link', closing: '}' },
  { opening: '{mat', closing: '}' },
  { opening: '{dxt', closing: '}' }
];

export const COMPLEX_TAGS_TO_REMOVE: ComplexTagType[] = [
  { opening: '{ds|', closing: '}' }
];

export const HW_TAGS: TagType[] = [
  { tag: '{bit}', replacement: '' },
  { tag: '{/bit}', replacement: '' }
];

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto'
  //   futr: 'Futuro indicativo',
  //   cond: 'Condicional',
  //   psub: 'Presente subjentivo',
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
