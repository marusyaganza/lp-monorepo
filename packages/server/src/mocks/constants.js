export const mockEnv = {
  SALT_NUMBER: 3,
  JWT_SECTET: 'JWT_secret'
};
