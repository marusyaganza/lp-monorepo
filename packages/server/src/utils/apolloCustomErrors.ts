import { GraphQLError } from 'graphql';
import { ApolloServerErrorCode } from '@apollo/server/errors';
import { ERROR_MESSAGES } from '../constants/errorMessages';

export enum ERROR_CODES {
  Unauthorized = 'UNAUTHORIZED',
  Forbidden = 'FORBISSEN'
}

export class AuthenticationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHENTICATED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Unauthorized } });
  }
}

export class AuthorizationError extends GraphQLError {
  constructor(errorMessage = ERROR_MESSAGES.NOT_AUTHORIZED) {
    super(errorMessage, { extensions: { code: ERROR_CODES.Forbidden } });
  }
}

export class UserInputError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: { code: ApolloServerErrorCode.BAD_USER_INPUT }
    });
  }
}

export class OperationResolutionError extends GraphQLError {
  constructor(errorMessage: string) {
    super(errorMessage, {
      extensions: {
        code: ApolloServerErrorCode.OPERATION_RESOLUTION_FAILURE
      }
    });
  }
}
