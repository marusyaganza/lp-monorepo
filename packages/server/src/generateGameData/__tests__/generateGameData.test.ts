import { generateGameData } from '..';
import { Game, Language, Tense } from '../../generated/graphql';
import {
  connectToDb,
  disconnectFromDb,
  dropDb,
  seedDb
} from '../../tests/helpers';

import { usersTestData } from '../../tests/mocks/inputs/newUserInput';
import { newWordInputs } from '../../tests/mocks/inputs/newWordInputs';
import { OperationResolutionError } from '../../utils/apolloCustomErrors';
import { ERROR_MESSAGES } from '../../constants/errorMessages';
import { DEFAULT_GAMES_SETTINGS } from '../../constants/defultGameSettings';

const snapshotConfig = {
  wordId: expect.any(String)
};

const gameTypes = Object.values(Game).filter(
  g => g !== Game.Conjugation && g !== Game.Gender
);
const tenses = Object.values(Tense);

describe('generateGameData', () => {
  let data;
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
    data = await seedDb({
      words: [
        ...newWordInputs[Language.Spanish],
        ...newWordInputs[Language.English]
      ],
      users: usersTestData
    });
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });
  gameTypes.forEach(game => {
    const gameConfig = DEFAULT_GAMES_SETTINGS[game];
    const wordsPerGame = gameConfig?.wordsPerGame;

    test(`with ${game} game Spanish`, async () => {
      const userId = data?.users?.[0] as string;
      const resultSpanish = await generateGameData(
        {
          gameType: game,
          language: Language.Spanish
        },
        userId
      );

      expect(resultSpanish?.questions?.[0]).toMatchSnapshot(snapshotConfig);
      expect(resultSpanish.questions).toHaveLength(wordsPerGame);
    });

    test(`with ${game} game English`, async () => {
      const userId = data?.users?.[0] as string;
      const result = await generateGameData(
        {
          gameType: game,
          language: Language.English
        },
        userId
      );

      expect(result?.questions?.[0]).toMatchSnapshot(snapshotConfig);
      expect(result.questions).toHaveLength(wordsPerGame);
    });

    test(`should throw an error with ${game} game if not enough words`, async () => {
      const userId = data?.users?.[1] as string;
      let error;
      try {
        await generateGameData(
          {
            gameType: game
          },
          userId
        );
      } catch (err) {
        error = err;
      }
      expect(error).toEqual(
        new OperationResolutionError(
          `not enough words to start a game. You have 0 word. Words requited for the game: ${gameConfig?.minWords}`
        )
      );
    });
  });

  tenses.forEach(tense => {
    test(`with ${Game.Conjugation} game ${tense} tense`, async () => {
      const userId = data?.users?.[0];
      const gameConfig = DEFAULT_GAMES_SETTINGS[Game.Conjugation];
      const wordsPerGame = gameConfig?.wordsPerGame;

      const result = await generateGameData(
        {
          gameType: Game.Conjugation,
          language: Language.Spanish,
          tense
        },
        userId
      );

      expect(result?.questions?.[0]).toMatchSnapshot(snapshotConfig);
      expect(result.questions).toHaveLength(wordsPerGame);
    });
  });

  test(`should throw error with ${Game.Conjugation} game ${Language.English}`, async () => {
    const userId = data?.users?.[0];
    let error;
    try {
      await generateGameData(
        {
          gameType: Game.Conjugation,
          language: Language.English
        },
        userId
      );
    } catch (err) {
      error = err;
    }
    expect(error).toEqual(
      new OperationResolutionError(ERROR_MESSAGES.GAME_NOT_AVAILABLE)
    );
  });

  test(`should throw error with ${Game.Gender} game ${Language.English}`, async () => {
    const userId = data?.users?.[0];
    let error;
    try {
      await generateGameData(
        {
          gameType: Game.Gender,
          language: Language.English
        },
        userId
      );
    } catch (err) {
      error = err;
    }
    expect(error).toEqual(
      new OperationResolutionError(ERROR_MESSAGES.GAME_NOT_AVAILABLE)
    );
  });
});
