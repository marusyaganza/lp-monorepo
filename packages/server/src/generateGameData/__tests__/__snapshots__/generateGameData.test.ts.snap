// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGameData with AUDIO game English 1`] = `
{
  "additionalInfo": {
    "examples": [
      {
        "text": "She found the perfect book by sheer serendipity.",
      },
    ],
    "imgUrl": "https://example.com/images/serendipity.jpg",
    "shortDef": "<b>serendipity means</b> Unexpected good fortune.",
  },
  "answer": "serendipity",
  "question": [
    "https://example.com/audio/serendipity.mp3",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with AUDIO game Spanish 1`] = `
{
  "additionalInfo": {
    "examples": [
      {
        "text": "Él puede correr muy rápido.",
        "translation": "He can run very fast.",
      },
    ],
    "imgUrl": "https://example.com/images/correr.jpg",
    "shortDef": "<b>correr means</b> To run.",
  },
  "alternativeSpelling": [
    "korrer",
  ],
  "answer": "correr",
  "question": [
    "https://example.com/audio/correr.mp3",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with CONJUGATION game impf tense 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "imgUrl": "https://example.com/images/correr.jpg",
    "shortDef": "<b>correr means</b> To run.",
  },
  "answer": "corre, corra, corramos, corred, corran",
  "question": [
    "correr",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with CONJUGATION game pind tense 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "imgUrl": "https://example.com/images/correr.jpg",
    "shortDef": "<b>correr means</b> To run.",
  },
  "answer": "corro, corres, corre, corremos, corréis, corren",
  "question": [
    "correr",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with CONJUGATION game pprf tense 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "imgUrl": "https://example.com/images/correr.jpg",
    "shortDef": "<b>correr means</b> To run.",
  },
  "answer": "he corrido, has corrido, ha corrido, hemos corrido, habéis corrido, han corrido",
  "question": [
    "correr",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with CONJUGATION game pret tense 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "imgUrl": "https://example.com/images/correr.jpg",
    "shortDef": "<b>correr means</b> To run.",
  },
  "answer": "corrí, corriste, corrió, corrimos, corristeis, corrieron",
  "question": [
    "correr",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with SELECT_DEF game English 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/serendipity.mp3",
    "examples": [
      {
        "text": "She found the perfect book by sheer serendipity.",
      },
    ],
    "imgUrl": "https://example.com/images/serendipity.jpg",
  },
  "answer": "Unexpected good fortune.",
  "options": [
    "Short-lived.",
    "Found everywhere.",
    "Idealistic but impractical.",
    "State of being alone.",
    "Glowing or bright.",
    "Unexpected good fortune.",
  ],
  "question": [
    "serendipity",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with SELECT_DEF game Spanish 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "examples": [
      {
        "text": "Él puede correr muy rápido.",
        "translation": "He can run very fast.",
      },
    ],
    "imgUrl": "https://example.com/images/correr.jpg",
  },
  "answer": "To run.",
  "options": [
    "Happiness.",
    "Quickly.",
    "Big, large.",
    "To think.",
    "Beauty.",
    "To run.",
  ],
  "question": [
    "correr",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with SELECT_WORD game English 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/serendipity.mp3",
    "examples": [
      {
        "text": "She found the perfect book by sheer serendipity.",
      },
    ],
    "imgUrl": "https://example.com/images/serendipity.jpg",
  },
  "answer": "serendipity",
  "options": [
    "ephemeral",
    "ubiquitous",
    "quixotic",
    "solitude",
    "luminous",
    "serendipity",
  ],
  "question": [
    "Unexpected good fortune.",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with SELECT_WORD game Spanish 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "examples": [
      {
        "text": "Él puede correr muy rápido.",
        "translation": "He can run very fast.",
      },
    ],
    "imgUrl": "https://example.com/images/correr.jpg",
  },
  "answer": "correr",
  "options": [
    "felicidad",
    "rápidamente",
    "grande",
    "pensar",
    "hermosura",
    "correr",
  ],
  "question": [
    "To run.",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with TYPE_WORD game English 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/serendipity.mp3",
    "examples": [
      {
        "text": "She found the perfect book by sheer serendipity.",
      },
    ],
    "imgUrl": "https://example.com/images/serendipity.jpg",
  },
  "answer": "serendipity",
  "question": [
    "Unexpected good fortune.",
  ],
  "wordId": Any<String>,
}
`;

exports[`generateGameData with TYPE_WORD game Spanish 1`] = `
{
  "additionalInfo": {
    "audioUrl": "https://example.com/audio/correr.mp3",
    "examples": [
      {
        "text": "Él puede correr muy rápido.",
        "translation": "He can run very fast.",
      },
    ],
    "imgUrl": "https://example.com/images/correr.jpg",
  },
  "alternativeSpelling": [
    "korrer",
  ],
  "answer": "correr",
  "question": [
    "To run.",
  ],
  "wordId": Any<String>,
}
`;
